crypto/addresses/src/bech32.rs

```diff
@@ -0,0 +1,161 @@
+    100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100,
+    100, 100, 15, 100, 10, 17, 21, 20, 26, 30, 7, 5, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100,
+    100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100,
+    100, 100, 100, 29, 100, 24, 13, 25, 9, 8, 23, 100, 18, 22, 31, 27, 19, 100, 1, 0, 3, 16, 11, 28, 12, 14, 6, 4, 2,
+];
+
+// Checksome for bech32
+// https://bch.info/en/specifications
+fn polymod<I>(values: I) -> u64
+where
+    I: Iterator<Item=u8>,
+{
+    let mut c = 1u64;
+    for d in values {
+        let c0 = c >> 35;
+        c = ((c & 0x07ffffffff) << 5) ^ (d as u64);
+
+        if c0 & 0x01 != 0 {
+            c ^= 0x98f2bc8e61;
+        }
+        if c0 & 0x02 != 0 {
+            c ^= 0x79b76d99e2;
+        }
+        if c0 & 0x04 != 0 {
+            c ^= 0xf33e5fb3c4;
+        }
+        if c0 & 0x08 != 0 {
+            c ^= 0xae2eabe2a8;
+        }
+        if c0 & 0x10 != 0 {
+            c ^= 0x1e4f43e470;
+        }
+    }
+    c ^ 1
+}
+
+fn checksum(payload: &[u8], prefix: &[u8]) -> u64 {
+    polymod(
+        prefix
+            .iter().copied()
+            .chain([0u8])
+            .chain(payload.iter().copied())
+            .chain([0u8; 8]),
+    )
+}
+
+// Convert 8bit array to 5bit array with right padding
+fn conv8to5 (payload: &[u8]) -> Vec<u8> {
+    let padding = match payload.len() % 5 == 0 {
+        true => 0,
+        false => 1
+    };
+    let mut five_bit = vec![0u8; payload.len() * 8 / 5 + padding];
+    let mut current_idx = 0;
+
+    let mut buff = 0u16;
+    let mut bits = 0;
+    for char in payload.iter() {
+        buff = (buff << 8) | *char as u16;
+        bits += 8;
+        while bits >= 5 {
+            bits -= 5;
+            five_bit[current_idx] = (buff >> bits) as u8;
+            buff = buff & ((1 << bits) - 1);
+            current_idx += 1;
+        }
+    }
+    if bits > 0 {
+        five_bit[current_idx] = (buff << (5-bits)) as u8;
+    }
+    five_bit
+}
+
+// Convert 5 bit array to 8 bit array, ignore right side padding
+fn conv5to8(payload: &[u8]) -> Vec<u8> {
+    let mut eight_bit = vec![0u8; payload.len() * 5 / 8];
+    let mut current_idx = 0;
+
+    let mut buff = 0u16;
+    let mut bits = 0;
+    for char in payload.iter() {
+        buff = (buff << 5) | *char as u16;
+        bits += 5;
+        while bits >= 8 {
+            bits -= 8;
+            eight_bit[current_idx] = (buff >> bits) as u8;
+            buff = buff & ((1 << bits) - 1);
+            current_idx += 1;
+        }
+    }
+    eight_bit
+}
+
+impl Address {
+    pub(crate) fn encode_payload(&self) -> String {
+        // Convert into 5 bits vector
+        let fivebit_payload = conv8to5(&[vec![self.version], self.payload.clone()].concat());
```

Comment:
This also doesn't require vec

Reply:
Should I use `as_slice` instead? to make them both `&[u8]`?

Reply:
Yeah

Reply:
done